sources/api/netdog/src/networkd_status.rs

```diff
@@ -0,0 +1,304 @@
+#[serde(deny_unknown_fields)]
+#[allow(dead_code)]
+pub(crate) struct SearchDomain {
+    #[serde(rename = "Domain")]
+    pub(crate) domain: String,
+    #[serde(rename = "ConfigSource")]
+    config_source: String,
+    #[serde(rename = "ConfigProvider", deserialize_with = "ipaddr_from_vec_de")]
+    config_provider: IpAddr,
+}
+
+#[derive(Debug, Deserialize)]
+pub(crate) struct NetworkctlIpAddr {}
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(rename_all = "PascalCase")]
+#[allow(dead_code)]
+pub(crate) struct NetworkdStatus {
+    pub(crate) name: InterfaceName,
+    #[serde(rename = "DNS")]
+    pub(crate) dns: Option<Vec<NetworkdDnsConfig>>,
+    pub(crate) search_domains: Option<Vec<SearchDomain>>,
+    #[serde(rename = "HardwareAddress")]
+    pub(crate) mac_address: Vec<u8>,
+    #[serde(rename = "Addresses", deserialize_with = "from_networkctl_addresses")]
+    pub(crate) addresses: Vec<IpAddr>,
+}
+
+// get an IpAddr from a Vec<u8> (could be 4 or 16 length)
+fn ipaddr_from_vec(address_vec: Vec<u8>) -> std::result::Result<IpAddr, Error> {
+    use error::BadIpAddressSnafu;
+    match address_vec.len() {
+        4 => {
+            let octets: [u8; 4] = address_vec.try_into().unwrap();
+            Ok(IpAddr::from(octets))
+        }
+        16 => {
+            let octets: [u8; 16] = address_vec.try_into().unwrap();
+            Ok(IpAddr::from(octets))
+        }
+        _ => BadIpAddressSnafu {
+            input: address_vec,
+            msg: "invalid length, must be 4 or 16 octets".to_string(),
+        }
+        .fail(),
+    }
+}
+
+fn ipaddr_from_vec_de<'de, D>(deserializer: D) -> std::result::Result<IpAddr, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    use serde::de::Error;
+    let address_vec: Vec<u8> = Deserialize::deserialize(deserializer)?;
+    ipaddr_from_vec(address_vec).map_err(D::Error::custom)
+}
+
+fn from_networkctl_addresses<'de, D>(deserializer: D) -> std::result::Result<Vec<IpAddr>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    use serde::de::Error;
+    #[derive(Debug, Deserialize)]
+    #[serde(rename_all = "PascalCase")]
+    #[allow(dead_code)]
+    struct NetworkctlAddress {
+        family: u8,
+        address: Vec<u8>,
+        prefix_length: u8,
+        config_source: String,
+    }
+    let addresses: Vec<NetworkctlAddress> = Deserialize::deserialize(deserializer)?;
+    let mut addrs = Vec::new();
+    for addr in addresses.iter() {
+        addrs.push(ipaddr_from_vec(addr.address.clone()).map_err(D::Error::custom)?);
+    }
+    Ok(addrs)
+}
+
+#[cfg(net_backend = "systemd-networkd")]
+impl NetworkdStatus {
+    pub(crate) fn primary_address(&self) -> Result<IpAddr> {
+        use error::NoIpAddressSnafu;
+        match self.addresses.len().cmp(&1) {
+            Ordering::Less => NoIpAddressSnafu {
+                interface: self.name.clone(),
+            }
+            .fail(),
+            Ordering::Equal => Ok(self.addresses[0]),
+            Ordering::Greater => {
```

Comment:
Can we add a comment here explaining this?

Reply:
Yeah, that is a good callout, this has some opinion about the "right" IP Address worth calling out.